### ISSUE TITLE
Hold Music for Attended Transfer

### ISSUE THREAD

**Issue by sipsorcery** [sipsorcery/sipsorcery#192]
> I've added a music on hold feature to the softphone. Now when you put a call on hold the remote party will hear music rather than still getting the microphone input.

how can it be archived to use the sip-server default hold music instead of the sample music?

_Originally posted by @blendR8 in https://github.com/sipsorcery/sipsorcery/issues/145#issuecomment-611564774_

**Comment by sipsorcery** [comment]
Do you mean play a stream from a different server to the on hold party when doing an attended transfer?

That would get pretty complicated to juggle the SIP calls. Here's a [SIP call flow diagram](https://sipsorcery.github.io/sipsorcery/articles/attendedtransfer.html) for a standard Attended Transfer. What you're talking about is introducing a 4th call leg to get a music on hold stream flowing.

It's definitely possible and if you control the music on hold server and SIP clients it could be reasonably reliable. But if you're talking about trying to support arbitrary SIP servers and clients then I don't think it would be very practical. The standard Attended Transfer with 3 legs often fails due to incompatibilities between different SIP stacks. Is it worth risking a transfer failure to get 3rd party on hold music?

**Comment by blendR8** [comment]
Not a different server. The sip-server where I am registered. If I use a hardware phone like snom or a softwarephone like phonerlite and I put a caller on hold, They here the same music. Therefore I think it is a sip server default hold music. The server is asterisk. Can you give me a hint?BTW. Your project is awesome. Thank you 😊 Anfang der weitergeleiteten E-Mail


Von: "Aaron Clauson" <notifications@github.com>Datum: 9. April 2020An: sipsorcery/sipsorcery <sipsorcery@noreply.github.com>Cc: Mention <mention@noreply.github.com>,blendR8 <sebastian.stefan@gmx.de>Betreff: Re: [sipsorcery/sipsorcery] Hold Music for Attended Transfer (#192)
Do you mean play a stream from a different server to the on hold party when doing an attended transfer?
That would get pretty complicated to juggle the SIP calls. Here's a SIP call flow diagram for a standard Attended Transfer. What you're talking about is introducing a 4th call leg to get a music on hold stream flowing.
It's definitely possible and if you control the music on hold server and SIP clients it could be reasonably reliable. But if you're talking about trying to support arbitrary SIP servers and clients then I don't think it would be very practical. The standard Attended Transfer with 3 legs often fails due to incompatibilities between different SIP stacks. Is it worth risking a transfer failure to get 3rd party on hold music?

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe.
[
{
"@context": "http://schema.org",
"@type": "EmailMessage",
"potentialAction": {
"@type": "ViewAction",
"target": "https://github.com/sipsorcery/sipsorcery/issues/192#issuecomment-611644332",
"url": "https://github.com/sipsorcery/sipsorcery/issues/192#issuecomment-611644332",
"name": "View Issue"
},
"description": "View this Issue on GitHub",
"publisher": {
"@type": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

**Comment by sipsorcery** [comment]
Coincidentally the music on hold file I used in the softphone is the default Asterisk music on hold.

It's straight forward to stop the softphone app from generating music when it places a call on hold. Change the audio source set [here](https://github.com/sipsorcery/sipsorcery/blob/5e6186e773b16fc6419352df8daa46654d5d1f9e/examples/Softphone/Signalling/SIPClient.cs#L257) to:

`AudioSource = AudioSourcesEnum.None`

**Comment by blendR8** [comment]
Stopping the music works, but doesn't do what i want. I have 2 files attached. both are logs from putting a caller in hold. 
[Phoner.txt](https://github.com/sipsorcery/sipsorcery/files/4481991/Phoner.txt)
[sipsorcery.txt](https://github.com/sipsorcery/sipsorcery/files/4481992/sipsorcery.txt)

the first is showing a log from PhonerLite
the second is showing a log from my app

I see that the log in my app is much longer. I also see that in my app the line "a=recvonly" is missing

**Comment by sipsorcery** [comment]
I've been doing some digging with softphones, Asterisk and music on hold in #197. I've been looking at a different issue that crops up but I don't think it's related to your problem or at least not the one from your SIP trace.

The first problem from your `sipsorcery.txt` file is the `ACK` request from your app isn't accepted by the Asterisk server hence the retransmits. One possible reason is that for some reason you've put an IP address in the `Call-ID` header, e.g. `Call-ID: 5ba6b9732f2bca870136e3b436660880@10.10.10.5:5060`. This is generally not a good idea in case there are routers, gateways that mangle IP addresses. The `Call-ID` is a critical header for matching SIP transactions and if it gets adjusted situations like mismatched ACK's occur.

**Comment by blendR8** [comment]
So what should I do or try? I see the same call-id header in the other softphone log where this problem doesn't appear. I also did't change the header manually. How can I change the header?

**Comment by sipsorcery** [comment]
It's a mystery as to what's setting the `Call-ID` header in your sipsorcery trace. The user agent indicates it's the sipsorcery library but there's nowhere in the code that the IP address gets appended...

`Call-ID: 5ba6b9732f2bca870136e3b436660880@10.10.10.5:5060`

If you're able you could do a full WireShark capture of the call to enable further diagnosis. If you don't want to attach it to a public forum you can email it to me at aaron@sipsorcery.com.

**Comment by blendR8** [comment]
can you tell me how to filter the interesting stuff from the WS protocol?

**Comment by sipsorcery** [comment]
> can you tell me how to filter the interesting stuff from the WS protocol?

Search for "wireshark sip capture" and you'll find lots of references.

**Comment by blendR8** [comment]
you got a mail.

**Comment by sipsorcery** [comment]
My tentative theory about the issue being related to SIP signalling looks like it was wrong. From your trace the re-INVITE SIP messages do look correct although there looks to be some kind of network or application delay causing retransmits.

This is from your trace:

![invite_moh](https://user-images.githubusercontent.com/197660/79747131-15acad00-8303-11ea-9571-b5799809ca60.png)

This is the same thing from the softphone to my Asterisk server all on a local network. Notice the absence of the Ok response and ACK retransmits:

![invite_moh_aaron](https://user-images.githubusercontent.com/197660/79747142-1b09f780-8303-11ea-8361-07e5e747fe5b.png)

There are two RTP streams from your trace:

![rtp_streams](https://user-images.githubusercontent.com/197660/79747386-7e942500-8303-11ea-932b-0773a51e5ae5.png)

Here is what I think is happening based on the trace:

 - You place a call from the softphone to an Asterisk server which I assume then places a call to an external SIP Provider (this call leg is not in the trace),
 - At the 55s mark you press the Hold button on the softphone,
 - The softphone stops sending RTP (which is most likely because you've set the audio source to None as recommended above) rather than music on hold,
 - Asterisk continues sending RTP to the softphone even though it's been asked not to (my Asterisk server does this as well which is weird but at least on my end could be an Asterisk misconfiguration). This isn't the problem though as all that means is the softphone is still getting audio from somewhere:

![rtp_stream](https://user-images.githubusercontent.com/197660/79748246-07f82700-8305-11ea-9742-3140a5eec496.png)

- At the 62s mark you press the button on the softphone to take the call off hold and the RTP from the softphone to Asterisk starts flowing again.

But unfortunately I don't think any of this has anything to do  with the issue you're facing. I'm guessing what you want is some custom music on hold to be played to the destination of the call leg from Asterisk to the SIP Provider. The softphone is not in control of that call leg. You'll need to troubleshoot your Asterisk configuration to see why it's not playing music to the remote party once the softphone has put the call on hold.

**Comment by sipsorcery** [comment]
After all that I had a second look at your `Phoner` capture and the I did spot a critical difference between your Asterisk server and mine.

This is the trace from my server:

````
INVITE sip:asterisk@192.168.11.48:5060 SIP/2.0
Via: SIP/2.0/UDP 192.168.11.50:5060;branch=z9hG4bK9af7a659f6a34fd68ebdf8bef1dc5362;rport
To: <sip:7001@192.168.11.48>;tag=b0bb2e1a-9922-49fa-ba3f-7495aa36219f
From: <sip:192.168.11.50>;tag=NRDUKXVMGO
Call-ID: f2c5cdfa-48a9-42c5-902e-5ad5f3f1d822
CSeq: 6052 INVITE
Contact: <sip:192.168.11.50:5060>
Max-Forwards: 70
User-Agent: www.sipsorcery.com
Supported: 100rel
Content-Length: 186
Content-Type: application/sdp

v=0
o=- 70280 0 IN IP4 127.0.0.1
s=-
c=IN IP4 192.168.11.50
t=0 0
m=audio 13930 RTP/AVP 101 0
a=rtpmap:101 telephone-event/8000
a=fmtp:101 0-16
a=rtpmap:0 PCMU/8000
a=sendonly

SIP/2.0 200 OK
Via: SIP/2.0/UDP 192.168.11.50:5060;rport=5060;received=192.168.11.50;branch=z9hG4bK9af7a659f6a34fd68ebdf8bef1dc5362
Call-ID: f2c5cdfa-48a9-42c5-902e-5ad5f3f1d822
From: <sip:192.168.11.50>;tag=NRDUKXVMGO
To: <sip:7001@192.168.11.48>;tag=b0bb2e1a-9922-49fa-ba3f-7495aa36219f
CSeq: 6052 INVITE
Contact: <sip:asterisk@192.168.11.48:5060>
Allow: OPTIONS, REGISTER, SUBSCRIBE, NOTIFY, PUBLISH, INVITE, ACK, BYE, CANCEL, UPDATE, PRACK, MESSAGE, REFER
Supported: 100rel, timer, replaces, norefersub
Server: Asterisk PBX 17.0.1
Content-Type: application/sdp
Content-Length:   238

v=0
o=- 1980339226 1980339228 IN IP4 192.168.11.50
s=Asterisk
c=IN IP4 192.168.11.50
t=0 0
m=audio 4002 RTP/AVP 0 101
a=rtpmap:0 PCMU/8000
a=rtpmap:101 telephone-event/8000
a=fmtp:101 0-16
a=ptime:20
a=maxptime:150
a=recvonly
````

This is the trace from your server:

````
INVITE sip:xxx@10.10.10.5:5060 SIP/2.0
Via: SIP/2.0/UDP 10.10.10.25:5060;branch=z9hG4bK8d9a4367f18d4a969643d30133b45f09;rport
To: "Durchwahl: xxx" <sip:xxx@10.10.10.5>;tag=as65a72548
From: <sip:10.10.10.25:5060>;tag=NZYFCFMMHT
Call-ID: 6719c2644c6842b4248ae3645b5529e6@10.10.10.5:5060
CSeq: 103 INVITE
Contact: <sip:10.10.10.25:5060>
Max-Forwards: 70
User-Agent: www.sipsorcery.com
Supported: 100rel
Content-Length: 184
Content-Type: application/sdp

v=0
o=- 13585 0 IN IP4 127.0.0.1
s=-
c=IN IP4 10.10.10.25
t=0 0
m=audio 11958 RTP/AVP 101 0
a=rtpmap:101 telephone-event/8000
a=fmtp:101 0-16
a=rtpmap:0 PCMU/8000
a=sendonly

SIP/2.0 200 OK
Via: SIP/2.0/UDP 10.10.10.25:5060;branch=z9hG4bK8d9a4367f18d4a969643d30133b45f09;received=10.10.10.25;rport=5060
From: <sip:10.10.10.25:5060>;tag=NZYFCFMMHT
To: "Durchwahl: xxx" <sip:xxx@10.10.10.5>;tag=as65a72548
Call-ID: 6719c2644c6842b4248ae3645b5529e6@10.10.10.5:5060
CSeq: 103 INVITE
Server: Asterisk PBX 11.3.0
Allow: INVITE, ACK, CANCEL, OPTIONS, BYE, REFER, SUBSCRIBE, NOTIFY, INFO, PUBLISH
Supported: replaces, timer
Contact: <sip:xxx@10.10.10.5:5060>
Content-Type: application/sdp
Content-Length: 173

v=0
o=root 980317513 980317513 IN IP4 10.10.10.5
s=Asterisk PBX 11.3.0
c=IN IP4 10.10.10.5
t=0 0
m=audio 10594 RTP/AVP 0
a=rtpmap:0 PCMU/8000
a=ptime:20
a=sendrecv
````

Look at the last line in both traces. The `sendrecv` RTP attribute is what dictates how the RTP stream will flow. In both cases when the softphone puts the call on hold it indicates so by setting `a=sendonly`. My Asterisk server agrees and responds with `a=recvonly`. Your server seems unhappy about something and leaves the RTP stream status at the original `a=sendrecv`.

The solution is still likely to be the same. Dig into your Asterisk music on hold settings.

**Comment by sipsorcery** [comment]
Perhaps you don't have the mulaw music on hold files downloaded for your Asterisk server?

That's another difference between the Phoner and sipsorcery softphone traces. Phoner has ALAW as the preferred codec the softphone only supports ULAW.

phoner:

````
m=audio 16960 RTP/AVP 9 18 8 0 101
a=rtpmap:9 G722/8000
a=rtpmap:18 G729/8000
a=fmtp:18 annexb=no
a=rtpmap:8 PCMA/8000
a=rtpmap:0 PCMU/8000
````

sipsorcery softphone:

````
m=audio 11434 RTP/AVP 101 0
a=rtpmap:101 telephone-event/8000
a=fmtp:101 0-16
a=rtpmap:0 PCMU/8000
````

**Comment by blendR8** [comment]
thank you for Investigation. I also noticed the missing recvonly in my post with the log-files. The problem is that i don't have direct access to the asterisk. Can you give me some information for changing the codec?

**Comment by sipsorcery** [comment]
In `SIPClient.cs` you can adjust the audio options supplied to the `RtpAVSession` constructor. Note that the only two audio codecs supported are PCMU and PCMA.

````
audioOpts = new AudioOptions
{
	AudioSource = AudioSourcesEnum.Microphone,
	OutputDeviceIndex = m_audioOutDeviceIndex,
	AudioCodecs = new List<SDPMediaFormatsEnum> {  SDPMediaFormatsEnum.PCMA, SDPMediaFormatsEnum.PCMU }
};
````

**Comment by blendR8** [comment]
I enabled both codecs now. Problem is the same. What I don't understand is why is phoner getting the response "recvonly" correctly with the same server?

**Comment by sipsorcery** [comment]
I'd like to know as well. Maybe that version of Asterisk doesn't like the RTP event format ID being first in the list. In the line below it should ideally by `0 101`. I'll adjust that in the next nuget package but even so it should work.

````
m=audio 11434 RTP/AVP 101 0
````

One quick test would be to disable all but either ulaw and/or alaw in your Phoner config and see if that makes a difference.

**Comment by blendR8** [comment]
I tested microSIP and enabled ulaw only. It worked like it should. These are the steps from putting caller in hold and after that off hold:

`
No. Time Source Destination Protocol Length Info
 492 21.879104 10.10.10.25 10.10.10.5 SIP/SDP 1062 Request: INVITE sip:
+491*******@10.10.10.5:5060, in-dialog |
Frame 492: 1062 bytes on wire (8496 bits), 1062 bytes captured (8496 bits)
Ethernet II, Src: 00:ff:f5:82:ac:98 (00:ff:f5:82:ac:98), Dst: 00:ff:f6:82:ac:98 (00:ff:f6:82:ac:98)
Internet Protocol Version 4, Src: 10.10.10.25, Dst: 10.10.10.5
User Datagram Protocol, Src Port: 63595, Dst Port: 5060
Session Initiation Protocol (INVITE)
 Request-Line: INVITE sip:+491*******@10.10.10.5:5060 SIP/2.0
 Message Header
 Via: SIP/2.0/UDP 10.10.10.25:63595;rport;branch=z9hG4bKPjc529d7ae0d0e49ecb811796ef7fd0de7
 Max-Forwards: 70
 From: <sip:tele339@10.10.10.25;ob>;tag=db9084dcaedd46149c4f7f0c1e4a522a
 To: "Durchwahl: +491*******" <sip:+491*******@10.10.10.5>;tag=as2bb510a1
 Contact: <sip:tele339@10.10.10.25:63595;ob>
 Call-ID: 6452844b4dd418e45c4db75e3b291a99@10.10.10.5:5060
 [Generated Call-ID: 6452844b4dd418e45c4db75e3b291a99@10.10.10.5:5060]
 CSeq: 26299 INVITE
 Allow: PRACK, INVITE, ACK, BYE, CANCEL, UPDATE, INFO, SUBSCRIBE, NOTIFY, REFER, MESSAGE, OPTIONS
 Supported: replaces, 100rel, timer, norefersub
 Session-Expires: 1800
 Min-SE: 90
 User-Agent: MicroSIP/3.19.29
 Content-Type: application/sdp
 Content-Length: 319
 Message Body
 Session Description Protocol
 Session Description Protocol Version (v): 0
 Owner/Creator, Session Id (o): - 3796444829 3796444831 IN IP4 192.168.178.43
 Session Name (s): pjmedia
 Bandwidth Information (b): AS:84
 Time Description, active time (t): 0 0
 Session Attribute (a): X-nat:0
 Media Description, name and address (m): audio 4000 RTP/AVP 0 101
 Connection Information (c): IN IP4 192.168.178.43
 Bandwidth Information (b): TIAS:64000
 Media Attribute (a): rtcp:4001 IN IP4 192.168.178.43
 Media Attribute (a): rtpmap:0 PCMU/8000
 Media Attribute (a): rtpmap:101 telephone-event/8000
 Media Attribute (a): fmtp:101 0-16
 Media Attribute (a): ssrc:2705443 cname:39b32d12074d4dc8
 Media Attribute (a): sendonly
 [Generated Call-ID: 6452844b4dd418e45c4db75e3b291a99@10.10.10.5:5060]
No. Time Source Destination Protocol Length Info
 497 21.916772 10.10.10.5 10.10.10.25 SIP 617 Status: 100 Trying |
Frame 497: 617 bytes on wire (4936 bits), 617 bytes captured (4936 bits)
Ethernet II, Src: 00:ff:f6:82:ac:98 (00:ff:f6:82:ac:98), Dst: 00:ff:f5:82:ac:98 (00:ff:f5:82:ac:98)
Internet Protocol Version 4, Src: 10.10.10.5, Dst: 10.10.10.25
User Datagram Protocol, Src Port: 5060, Dst Port: 63595
Session Initiation Protocol (100)
 Status-Line: SIP/2.0 100 Trying
 Message Header
 Via: SIP/2.0/UDP 10.10.10.25:63595;branch=z9hG4bKPjc529d7ae0d0e49ecb811796ef7fd0de7;received=10.10.10.25;rport=63595
 From: <sip:tele339@10.10.10.25;ob>;tag=db9084dcaedd46149c4f7f0c1e4a522a
 To: "Durchwahl: +491*******" <sip:+491*******@10.10.10.5>;tag=as2bb510a1
 Call-ID: 6452844b4dd418e45c4db75e3b291a99@10.10.10.5:5060
 [Generated Call-ID: 6452844b4dd418e45c4db75e3b291a99@10.10.10.5:5060]
 CSeq: 26299 INVITE
 Server: Asterisk PBX 11.3.0
 Allow: INVITE, ACK, CANCEL, OPTIONS, BYE, REFER, SUBSCRIBE, NOTIFY, INFO, PUBLISH
 Supported: replaces, timer
 Contact: <sip:+491*******@10.10.10.5:5060>
 Content-Length: 0
No. Time Source Destination Protocol Length Info
 498 21.916834 10.10.10.5 10.10.10.25 SIP/SDP 877 Status: 200 OK |
Frame 498: 877 bytes on wire (7016 bits), 877 bytes captured (7016 bits)
Ethernet II, Src: 00:ff:f6:82:ac:98 (00:ff:f6:82:ac:98), Dst: 00:ff:f5:82:ac:98 (00:ff:f5:82:ac:98)
Internet Protocol Version 4, Src: 10.10.10.5, Dst: 10.10.10.25
User Datagram Protocol, Src Port: 5060, Dst Port: 63595
Session Initiation Protocol (200)
 Status-Line: SIP/2.0 200 OK
 Message Header
 Via: SIP/2.0/UDP 10.10.10.25:63595;branch=z9hG4bKPjc529d7ae0d0e49ecb811796ef7fd0de7;received=10.10.10.25;rport=63595
 From: <sip:tele339@10.10.10.25;ob>;tag=db9084dcaedd46149c4f7f0c1e4a522a
 To: "Durchwahl: +491*******" <sip:+491*******@10.10.10.5>;tag=as2bb510a1
 Call-ID: 6452844b4dd418e45c4db75e3b291a99@10.10.10.5:5060
 [Generated Call-ID: 6452844b4dd418e45c4db75e3b291a99@10.10.10.5:5060]
 CSeq: 26299 INVITE
 Server: Asterisk PBX 11.3.0
 Allow: INVITE, ACK, CANCEL, OPTIONS, BYE, REFER, SUBSCRIBE, NOTIFY, INFO, PUBLISH
 Supported: replaces, timer
 Contact: <sip:+491*******@10.10.10.5:5060>
 Content-Type: application/sdp
 Content-Length: 231
 Message Body
 Session Description Protocol
 Session Description Protocol Version (v): 0
 Owner/Creator, Session Id (o): root 1941155728 1941155729 IN IP4 10.10.10.5
 Session Name (s): Asterisk PBX 11.3.0
 Connection Information (c): IN IP4 10.10.10.5
 Time Description, active time (t): 0 0
 Media Description, name and address (m): audio 14382 RTP/AVP 0 101
 Media Attribute (a): rtpmap:0 PCMU/8000
 Media Attribute (a): rtpmap:101 telephone-event/8000
 Media Attribute (a): fmtp:101 0-16
 Media Attribute (a): ptime:20
 Media Attribute (a): recvonly
 [Generated Call-ID: 6452844b4dd418e45c4db75e3b291a99@10.10.10.5:5060]
No. Time Source Destination Protocol Length Info
 502 21.917216 10.10.10.25 10.10.10.5 SIP 447 Request: ACK sip:
+491*******@10.10.10.5:5060 |
Frame 502: 447 bytes on wire (3576 bits), 447 bytes captured (3576 bits)
Ethernet II, Src: 00:ff:f5:82:ac:98 (00:ff:f5:82:ac:98), Dst: 00:ff:f6:82:ac:98 (00:ff:f6:82:ac:98)
Internet Protocol Version 4, Src: 10.10.10.25, Dst: 10.10.10.5
User Datagram Protocol, Src Port: 63595, Dst Port: 5060
Session Initiation Protocol (ACK)
 Request-Line: ACK sip:+491*******@10.10.10.5:5060 SIP/2.0
 Message Header
 Via: SIP/2.0/UDP 10.10.10.25:63595;rport;branch=z9hG4bKPj7fcabcc219894493a5517a5d156fec71
 Max-Forwards: 70
 From: <sip:tele339@10.10.10.25;ob>;tag=db9084dcaedd46149c4f7f0c1e4a522a
 To: "Durchwahl: +491*******" <sip:+491*******@10.10.10.5>;tag=as2bb510a1
 Call-ID: 6452844b4dd418e45c4db75e3b291a99@10.10.10.5:5060
 [Generated Call-ID: 6452844b4dd418e45c4db75e3b291a99@10.10.10.5:5060]
 CSeq: 26299 ACK
 Content-Length: 0
No. Time Source Destination Protocol Length Info
 1308 29.957144 10.10.10.25 10.10.10.5 SIP/SDP 1062 Request: INVITE sip:
+491*******@10.10.10.5:5060, in-dialog |
Frame 1308: 1062 bytes on wire (8496 bits), 1062 bytes captured (8496 bits)
Ethernet II, Src: 00:ff:f5:82:ac:98 (00:ff:f5:82:ac:98), Dst: 00:ff:f6:82:ac:98 (00:ff:f6:82:ac:98)
Internet Protocol Version 4, Src: 10.10.10.25, Dst: 10.10.10.5
User Datagram Protocol, Src Port: 63595, Dst Port: 5060
Session Initiation Protocol (INVITE)
 Request-Line: INVITE sip:+491*******@10.10.10.5:5060 SIP/2.0
 Message Header
 Via: SIP/2.0/UDP 10.10.10.25:63595;rport;branch=z9hG4bKPjdf6d4be705924063848e8dc5c2f34335
 Max-Forwards: 70
 From: <sip:tele339@10.10.10.25;ob>;tag=db9084dcaedd46149c4f7f0c1e4a522a
 To: "Durchwahl: +491*******" <sip:+491*******@10.10.10.5>;tag=as2bb510a1
 Contact: <sip:tele339@10.10.10.25:63595;ob>
 Call-ID: 6452844b4dd418e45c4db75e3b291a99@10.10.10.5:5060
 [Generated Call-ID: 6452844b4dd418e45c4db75e3b291a99@10.10.10.5:5060]
 CSeq: 26300 INVITE
 Allow: PRACK, INVITE, ACK, BYE, CANCEL, UPDATE, INFO, SUBSCRIBE, NOTIFY, REFER, MESSAGE, OPTIONS
 Supported: replaces, 100rel, timer, norefersub
 Session-Expires: 1800
 Min-SE: 90
 User-Agent: MicroSIP/3.19.29
 Content-Type: application/sdp
 Content-Length: 319
 Message Body
 Session Description Protocol
 Session Description Protocol Version (v): 0
 Owner/Creator, Session Id (o): - 3796444829 3796444832 IN IP4 192.168.178.43
 Session Name (s): pjmedia
 Bandwidth Information (b): AS:84
 Time Description, active time (t): 0 0
 Session Attribute (a): X-nat:0
 Media Description, name and address (m): audio 4000 RTP/AVP 0 101
 Connection Information (c): IN IP4 192.168.178.43
 Bandwidth Information (b): TIAS:64000
 Media Attribute (a): rtcp:4001 IN IP4 192.168.178.43
 Media Attribute (a): sendrecv
 Media Attribute (a): rtpmap:0 PCMU/8000
 Media Attribute (a): rtpmap:101 telephone-event/8000
 Media Attribute (a): fmtp:101 0-16
 Media Attribute (a): ssrc:2705443 cname:39b32d12074d4dc8
 [Generated Call-ID: 6452844b4dd418e45c4db75e3b291a99@10.10.10.5:5060]
No. Time Source Destination Protocol Length Info
 1313 29.995441 10.10.10.5 10.10.10.25 SIP 617 Status: 100 Trying |
Frame 1313: 617 bytes on wire (4936 bits), 617 bytes captured (4936 bits)
Ethernet II, Src: 00:ff:f6:82:ac:98 (00:ff:f6:82:ac:98), Dst: 00:ff:f5:82:ac:98 (00:ff:f5:82:ac:98)
Internet Protocol Version 4, Src: 10.10.10.5, Dst: 10.10.10.25
User Datagram Protocol, Src Port: 5060, Dst Port: 63595
Session Initiation Protocol (100)
 Status-Line: SIP/2.0 100 Trying
 Message Header
 Via: SIP/2.0/UDP 10.10.10.25:63595;branch=z9hG4bKPjdf6d4be705924063848e8dc5c2f34335;received=10.10.10.25;rport=63595
 From: <sip:tele339@10.10.10.25;ob>;tag=db9084dcaedd46149c4f7f0c1e4a522a
 To: "Durchwahl: +491*******" <sip:+491*******@10.10.10.5>;tag=as2bb510a1
 Call-ID: 6452844b4dd418e45c4db75e3b291a99@10.10.10.5:5060
 [Generated Call-ID: 6452844b4dd418e45c4db75e3b291a99@10.10.10.5:5060]
 CSeq: 26300 INVITE
 Server: Asterisk PBX 11.3.0
 Allow: INVITE, ACK, CANCEL, OPTIONS, BYE, REFER, SUBSCRIBE, NOTIFY, INFO, PUBLISH
 Supported: replaces, timer
 Contact: <sip:+491*******@10.10.10.5:5060>
 Content-Length: 0
No. Time Source Destination Protocol Length Info
 1314 29.995498 10.10.10.5 10.10.10.25 SIP/SDP 877 Status: 200 OK |
Frame 1314: 877 bytes on wire (7016 bits), 877 bytes captured (7016 bits)
Ethernet II, Src: 00:ff:f6:82:ac:98 (00:ff:f6:82:ac:98), Dst: 00:ff:f5:82:ac:98 (00:ff:f5:82:ac:98)
Internet Protocol Version 4, Src: 10.10.10.5, Dst: 10.10.10.25
User Datagram Protocol, Src Port: 5060, Dst Port: 63595
Session Initiation Protocol (200)
 Status-Line: SIP/2.0 200 OK
 Message Header
 Via: SIP/2.0/UDP 10.10.10.25:63595;branch=z9hG4bKPjdf6d4be705924063848e8dc5c2f34335;received=10.10.10.25;rport=63595
 From: <sip:tele339@10.10.10.25;ob>;tag=db9084dcaedd46149c4f7f0c1e4a522a
 To: "Durchwahl: +491*******" <sip:+491*******@10.10.10.5>;tag=as2bb510a1
 Call-ID: 6452844b4dd418e45c4db75e3b291a99@10.10.10.5:5060
 [Generated Call-ID: 6452844b4dd418e45c4db75e3b291a99@10.10.10.5:5060]
 CSeq: 26300 INVITE
 Server: Asterisk PBX 11.3.0
 Allow: INVITE, ACK, CANCEL, OPTIONS, BYE, REFER, SUBSCRIBE, NOTIFY, INFO, PUBLISH
 Supported: replaces, timer
 Contact: <sip:+491*******@10.10.10.5:5060>
 Content-Type: application/sdp
 Content-Length: 231
 Message Body
 Session Description Protocol
 Session Description Protocol Version (v): 0
 Owner/Creator, Session Id (o): root 1941155728 1941155730 IN IP4 10.10.10.5
 Session Name (s): Asterisk PBX 11.3.0
 Connection Information (c): IN IP4 10.10.10.5
 Time Description, active time (t): 0 0
 Media Description, name and address (m): audio 14382 RTP/AVP 0 101
 Media Attribute (a): rtpmap:0 PCMU/8000
 Media Attribute (a): rtpmap:101 telephone-event/8000
 Media Attribute (a): fmtp:101 0-16
 Media Attribute (a): ptime:20
 Media Attribute (a): sendrecv
 [Generated Call-ID: 6452844b4dd418e45c4db75e3b291a99@10.10.10.5:5060]
`

**Comment by sipsorcery** [comment]
I've made that small tweak in the order of the codecs in the SDP description and its available in nuget package v4.0.42-pre. See if that makes any difference.

**Comment by blendR8** [comment]
![Unbenannt](https://user-images.githubusercontent.com/50368850/79947995-47dd1c80-8473-11ea-9693-c871379275e4.PNG)
![Unbenannt2](https://user-images.githubusercontent.com/50368850/79947998-490e4980-8473-11ea-8dfa-5edf763072ba.PNG)

the problem still exists after going to v4.0.42. The first screenshot shows the invite request, the second the wrong answer from the server

**Comment by blendR8** [comment]
another difference i see is that in other softphones the user is set in the from:

 From: <sip:tele339@10.10.10.25;ob>;tag=db9084dcaedd46149c4f7f0c1e4a522a

Perhaps also missing allow and supported lines do the trick?

 Allow: INVITE, ACK, CANCEL, OPTIONS, BYE, REFER, SUBSCRIBE, NOTIFY, INFO, PUBLISH
 Supported: replaces, timer

**Comment by sipsorcery** [comment]
After spending a good part of the day installing Asterisk v11.3.0 (tried old gcc versions on ubuntu 18, no go, installed new vm with ubuntu 16 same problem, installed new vm with ubuntu 14 success) I was finally able to replicate the issue.

Incidentally Asterisk v11 was [end-of-lifed](https://wiki.asterisk.org/wiki/display/AST/Asterisk+Versions) two years ago.

The issue turned out not to be related to codecs but was instead due to me forgetting to bump the SDP announcement version number when the stream status changed. I don't think many SIP/WebRTC stacks even rely on the SDP announcement version number anymore. But Asterisk v11.3.0 does. I've fixed the problem and pushed nuget package v4.0.43-pre.

````
[16:13:10 DBG] INVITE sip:7001@192.168.11.24:5060 SIP/2.0
Via: SIP/2.0/UDP 192.168.11.50:5060;branch=z9hG4bK10514cdad72245d887b2b58eecc3732e;rport
To: <sip:7001@192.168.11.24:5060>;tag=as2995589a
From: <sip:thisis@anonymous.invalid>;tag=NISXHRQDNV
Call-ID: 022749dbfd3c42a7a520c2e7f421407c
CSeq: 2 INVITE
Contact: <sip:7000@192.168.11.50:5060>
Max-Forwards: 70
User-Agent: www.sipsorcery.com
Supported: 100rel
Content-Length: 191
Content-Type: application/sdp

v=0
o=- 1188126966 1 IN IP4 127.0.0.1
s=-
c=IN IP4 192.168.11.50
t=0 0
m=audio 10916 RTP/AVP 0 101
a=rtpmap:0 PCMU/8000
a=rtpmap:101 telephone-event/8000
a=fmtp:101 0-16
a=sendonly

[16:13:10 DBG] Response received: udp:192.168.11.50:5060<-udp:192.168.11.24:5060
[16:13:10 DBG] SIP/2.0 100 Trying
Via: SIP/2.0/UDP 192.168.11.50:5060;branch=z9hG4bK10514cdad72245d887b2b58eecc3732e;received=192.168.11.50;rport=5060
To: <sip:7001@192.168.11.24:5060>;tag=as2995589a
From: <sip:thisis@anonymous.invalid>;tag=NISXHRQDNV
Call-ID: 022749dbfd3c42a7a520c2e7f421407c
CSeq: 2 INVITE
Contact: <sip:7001@192.168.11.24:5060>
Allow: INVITE, ACK, CANCEL, OPTIONS, BYE, REFER, SUBSCRIBE, NOTIFY, INFO, PUBLISH
Supported: replaces, timer
Content-Length: 0
Server: Asterisk PBX 11.3.0


[16:13:10 DBG] Response received: udp:192.168.11.50:5060<-udp:192.168.11.24:5060
[16:13:10 DBG] SIP/2.0 200 OK
Via: SIP/2.0/UDP 192.168.11.50:5060;branch=z9hG4bK10514cdad72245d887b2b58eecc3732e;received=192.168.11.50;rport=5060
To: <sip:7001@192.168.11.24:5060>;tag=as2995589a
From: <sip:thisis@anonymous.invalid>;tag=NISXHRQDNV
Call-ID: 022749dbfd3c42a7a520c2e7f421407c
CSeq: 2 INVITE
Contact: <sip:7001@192.168.11.24:5060>
Allow: INVITE, ACK, CANCEL, OPTIONS, BYE, REFER, SUBSCRIBE, NOTIFY, INFO, PUBLISH
Supported: replaces, timer
Content-Length: 264
Content-Type: application/sdp
Server: Asterisk PBX 11.3.0

v=0
o=root 1998887747 1998887748 IN IP4 192.168.11.24
s=Asterisk PBX 11.3.0
c=IN IP4 192.168.11.24
t=0 0
m=audio 18550 RTP/AVP 0 101
a=rtpmap:0 PCMU/8000
a=rtpmap:101 telephone-event/8000
a=fmtp:101 0-16
a=silenceSupp:off - - - -
a=ptime:20
a=recvonly
````

**Comment by blendR8** [comment]
You can Close this issue now. It works like it should. 

Thank you so much for your Investigation in this.
